File: package/ts-replay-parser/src/transform/uint.ts
/**
 * Fixed-width unsigned integer arithmetic helpers.
 *
 * The C# source relies on `uint` (32-bit) and `ulong` (64-bit) wrapping
 * semantics. JavaScript `number` cannot represent 64-bit integers exactly,
 * and its bitwise operators coerce to *signed* 32-bit, so we implement the
 * 64-bit paths with `BigInt` and keep 32-bit math explicit and masked.
 */

export const U32_MASK = 0xffffffff;
export const U64_MASK = 0xffffffffffffffffn;

/** Coerce to an unsigned 32-bit value (mirrors C# `(uint)`). */
export function u32(value: number): number {
  return value >>> 0;
}

/** Coerce a bigint to an unsigned 64-bit value (mirrors C# `(ulong)`). */
export function u64(value: bigint): bigint {
  return value & U64_MASK;
}

/** Truncate to the low 8 bits (mirrors C# `(byte)`). */
export function u8(value: number | bigint): number {
  return Number(BigInt(value) & 0xffn);
}

/** 32-bit unsigned multiply with wraparound (mirrors C# `uint * uint`). */
export function mul32(a: number, b: number): number {
  // Use BigInt to avoid float precision loss above 2^53, then mask.
  return Number((BigInt(a >>> 0) * BigInt(b >>> 0)) & BigInt(U32_MASK));
}

/** 64-bit unsigned multiply with wraparound. */
export function mul64(a: bigint, b: bigint): bigint {
  return u64(a * b);
}

/** 64-bit unsigned add with wraparound. */
export function add64(a: bigint, b: bigint): bigint {
  return u64(a + b);
}

/** 64-bit unsigned subtract with wraparound. */
export function sub64(a: bigint, b: bigint): bigint {
  return u64(a - b);
}

/** 32-bit unsigned add with wraparound. */
export function add32(a: number, b: number): number {
  return (a + b) >>> 0;
}

/** 32-bit unsigned subtract with wraparound. */
export function sub32(a: number, b: number): number {
  return (a - b) >>> 0;
}

/** Rotate a 64-bit value right by `count` bits. */
export function rotr64(value: bigint, count: number): bigint {
  const n = BigInt(count & 63);
  if (n === 0n) return u64(value);
  return u64((value >> n) | (value << (64n - n)));
}

/** Rotate a 64-bit value left by `count` bits. */
export function rotl64(value: bigint, count: number): bigint {
  const n = BigInt(count & 63);
  if (n === 0n) return u64(value);
  return u64((value << n) | (value >> (64n - n)));
}

/** Rotate a 32-bit value right by `count` bits. */
export function rotr32(value: number, count: number): number {
  const n = count & 31;
  if (n === 0) return value >>> 0;
  return (((value >>> n) | (value << (32 - n))) >>> 0);
}

/** Rotate a 32-bit value left by `count` bits. */
export function rotl32(value: number, count: number): number {
  const n = count & 31;
  if (n === 0) return value >>> 0;
  return (((value << n) | (value >>> (32 - n))) >>> 0);
}

/** Rotate an 8-bit value right by `count` bits. */
export function rotr8(value: number, count: number): number {
  const v = value & 0xff;
  const n = count & 7;
  if (n === 0) return v;
  return ((v >>> n) | (v << (8 - n))) & 0xff;
}

File: src-tauri/src/replay/transform/uint.rs
//! Fixed-width unsigned integer arithmetic helpers.
//!
//! Port of `package/ts-replay-parser/src/transform/uint.ts`. The TS source
//! emulates C# `uint`/`ulong` wraparound semantics with `BigInt` masking;
//! Rust's native `u32`/`u64` already have exact wrapping arithmetic, so this
//! is a direct simplification, not a bignum port.

pub fn u32(value: u32) -> u32 {
    value
}

pub fn u64(value: u64) -> u64 {
    value
}

pub fn u8(value: u64) -> u8 {
    (value & 0xff) as u8
}

pub fn mul32(a: u32, b: u32) -> u32 {
    a.wrapping_mul(b)
}

pub fn mul64(a: u64, b: u64) -> u64 {
    a.wrapping_mul(b)
}

pub fn add64(a: u64, b: u64) -> u64 {
    a.wrapping_add(b)
}

pub fn sub64(a: u64, b: u64) -> u64 {
    a.wrapping_sub(b)
}

pub fn add32(a: u32, b: u32) -> u32 {
    a.wrapping_add(b)
}

pub fn sub32(a: u32, b: u32) -> u32 {
    a.wrapping_sub(b)
}

/// Rotate a 64-bit value right by `count` bits (mirrors TS `rotr64`, which
/// masks `count & 63` before rotating and returns the value unchanged when
/// the masked count is 0).
pub fn rotr64(value: u64, count: u32) -> u64 {
    let n = count & 63;
    if n == 0 {
        return value;
    }
    value.rotate_right(n)
}

pub fn rotl64(value: u64, count: u32) -> u64 {
    let n = count & 63;
    if n == 0 {
        return value;
    }
    value.rotate_left(n)
}

pub fn rotr32(value: u32, count: u32) -> u32 {
    let n = count & 31;
    if n == 0 {
        return value;
    }
    value.rotate_right(n)
}

pub fn rotl32(value: u32, count: u32) -> u32 {
    let n = count & 31;
    if n == 0 {
        return value;
    }
    value.rotate_left(n)
}

/// Rotate an 8-bit value right by `count` bits (mirrors TS `rotr8`, which
/// masks the input to 8 bits and `count & 7` before rotating).
pub fn rotr8(value: u8, count: u32) -> u8 {
    let n = (count & 7) as u32;
    if n == 0 {
        return value;
    }
    value.rotate_right(n)
}

/// Rotate an 8-bit value left by `count` bits (mirrors the C# reference's
/// `RotateLeft(byte, int)` overload, added for the 13.x transforms).
pub fn rotl8(value: u8, count: u32) -> u8 {
    let n = count & 7;
    if n == 0 {
        return value;
    }
    value.rotate_left(n)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rotr64_zero_count_is_identity() {
        assert_eq!(rotr64(0x1234, 0), 0x1234);
        assert_eq!(rotr64(0x1234, 64), 0x1234);
    }

    #[test]
    fn rotr8_masks_to_byte() {
        assert_eq!(rotr8(0b1000_0001, 1), 0b1100_0000);
    }
}
